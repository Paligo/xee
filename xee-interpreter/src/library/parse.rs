// https://www.w3.org/TR/xpath-functions-31/#parsing-and-serializing

use xee_xpath_macros::xpath_fn;

use crate::{context, error, interpreter::Interpreter, sequence, wrap_xpath_fn};

use super::StaticFunctionDescription;

#[xpath_fn("fn:parse-xml($arg as xs:string?) as document-node(element(*))?")]
fn parse_xml(
    context: &context::DynamicContext,
    interpreter: &mut Interpreter,
    arg: Option<&str>,
) -> error::Result<Option<xot::Node>> {
    if let Some(arg) = arg {
        let documents = context.documents();
        let handle = documents
            .borrow_mut()
            .add_string(interpreter.xot_mut(), None, arg)
            .map_err(|_| error::Error::FODC0006)?;
        let doc = documents
            .borrow()
            .get_node_by_handle(handle)
            .ok_or(error::Error::FODC0006)?;
        Ok(Some(doc))
    } else {
        Ok(None)
    }
}

#[xpath_fn("fn:parse-xml-fragment($arg as xs:string?) as document-node()?")]
fn parse_xml_fragment(
    context: &context::DynamicContext,
    interpreter: &mut Interpreter,
    arg: Option<&str>,
) -> error::Result<Option<xot::Node>> {
    if let Some(arg) = arg {
        let documents = context.documents();
        let handle = documents
            .borrow_mut()
            .add_fragment_string(interpreter.xot_mut(), arg)
            .map_err(|_| error::Error::FODC0006)?;
        let doc = documents
            .borrow()
            .get_node_by_handle(handle)
            .ok_or(error::Error::FODC0006)?;
        Ok(Some(doc))
    } else {
        Ok(None)
    }
}

#[xpath_fn("fn:serialize($arg as item()*) as xs:string")]
fn serialize1(interpreter: &mut Interpreter, arg: &sequence::Sequence) -> error::Result<String> {
    let node = arg.normalize(" ", interpreter.xot_mut())?;
    Ok(interpreter.xot().to_string(node)?)
}

#[xpath_fn("fn:serialize($arg as item()*, $params as item()?) as xs:string")]
fn serialize2(
    interpreter: &mut Interpreter,
    arg: &sequence::Sequence,
    _params: Option<sequence::Item>,
) -> error::Result<String> {
    // TODO: nothing is done with the parameters yet
    let node = arg.normalize(" ", interpreter.xot_mut())?;
    Ok(interpreter.xot().to_string(node)?)
}

pub(crate) fn static_function_descriptions() -> Vec<StaticFunctionDescription> {
    vec![
        wrap_xpath_fn!(parse_xml),
        wrap_xpath_fn!(parse_xml_fragment),
        wrap_xpath_fn!(serialize1),
        wrap_xpath_fn!(serialize2),
    ]
}
