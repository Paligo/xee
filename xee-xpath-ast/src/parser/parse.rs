// parse XPath via pest

// XXX extra-grammatical constraints still need to be handled:
// https://www.w3.org/TR/xpath-31/#extra-grammatical-constraints
// XXX reserved function names still need to be handled:
// https://www.w3.org/TR/xpath-31/#id-reserved-fn-names

#[derive(Parser)]
#[grammar = "parser/xpath-31.pest"]
pub(crate) struct XPathParser;

#[cfg(test)]
mod tests {
    use super::*;
    use pest::iterators::Pairs;
    use pest::Parser;

    // parse function signature as described in
    // https://www.w3.org/TR/xpath-functions-31/#func-signatures
    // This is almost the same as an inline function definition except
    // for the name part.
    #[allow(clippy::result_large_err)]
    pub(crate) fn parse_signature(
        signature: &str,
    ) -> Result<Pairs<Rule>, pest::error::Error<Rule>> {
        XPathParser::parse(Rule::Signature, signature)
    }

    #[test]
    fn test_char() {
        let successful_parse = XPathParser::parse(Rule::Char, "abc");
        assert!(successful_parse.is_ok());
        let unsuccessful_parse = XPathParser::parse(Rule::Char, "\u{10}");
        assert!(unsuccessful_parse.is_err());
    }

    #[test]
    fn test_not_xpath() {
        let unsuccessful_parse = XPathParser::parse(Rule::Xpath, ")(");
        assert!(unsuccessful_parse.is_err());
    }

    #[test]
    fn test_not_xpath_incomplete() {
        let unsuccessful_parse = XPathParser::parse(Rule::Xpath, "foo(");
        assert!(unsuccessful_parse.is_err());
    }

    #[test]
    fn test_xpath_simple_name() {
        let successful_parse = XPathParser::parse(Rule::Xpath, "foo");
        assert!(successful_parse.is_ok());
    }

    #[test]
    fn test_xpath_relative_path() {
        let successful_parse = XPathParser::parse(Rule::Xpath, "foo/bar");
        assert!(successful_parse.is_ok());
    }

    #[test]
    fn test_xpath_absolute_path() {
        let successful_parse = XPathParser::parse(Rule::Xpath, "/foo/bar");
        assert!(successful_parse.is_ok());
    }

    #[test]
    fn test_signature() {
        let successful_parse = parse_signature("math:exp($arg as xs:double?) as xs:double?");
        assert!(successful_parse.is_ok());
    }

    #[test]
    fn test_signature_parameter_type_required() {
        let successful_parse = parse_signature("math:exp($arg) as xs:double?");
        assert!(successful_parse.is_err());
    }

    #[test]
    fn test_signature_return_type_required() {
        let successful_parse = parse_signature("math:exp($arg as xs:double?)");
        assert!(successful_parse.is_err());
    }
}
