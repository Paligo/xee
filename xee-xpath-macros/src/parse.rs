use syn::{
    parse::{Parse, ParseStream},
    token::Comma,
    LitStr, Result,
};

use xee_xpath_ast::ast::Signature;
use xee_xpath_ast::parse_signature;
use xee_xpath_ast::Namespaces;

#[derive(Debug)]
pub(crate) struct XPathFnOptions {
    pub(crate) signature: Signature,
    pub(crate) kind: Option<String>,
    pub(crate) signature_string: String,
}

mod kw {
    syn::custom_keyword!(context_first);
    syn::custom_keyword!(context_last);
    syn::custom_keyword!(position);
    syn::custom_keyword!(size);
}

impl Parse for XPathFnOptions {
    fn parse(input: ParseStream) -> Result<Self> {
        let mut options = Vec::new();
        while !input.is_empty() {
            options.push(input.parse()?);
            if !input.is_empty() {
                let _: Comma = input.parse()?;
            }
        }

        let mut signature: Option<String> = None;
        let mut kind: Option<String> = None;
        for option in options {
            match option {
                XPathFnOption::Signature(signature_option) => {
                    signature = Some(signature_option);
                }
                XPathFnOption::Kind(kind_option) => {
                    kind = Some(kind_option);
                }
            }
        }
        let signature_string = signature.unwrap();
        let namespaces = Namespaces::default();
        let signature = parse_signature(&signature_string, &namespaces)
            .map_err(|e| syn::Error::new(input.span(), format!("{:?}", e)))?;
        Ok(Self {
            signature,
            kind,
            signature_string,
        })
    }
}

enum XPathFnOption {
    Signature(String),
    Kind(String),
}

impl Parse for XPathFnOption {
    fn parse(input: ParseStream) -> Result<Self> {
        let lookahead = input.lookahead1();
        Ok(if lookahead.peek(kw::context_first) {
            let _eat: kw::context_first = input.parse()?;
            XPathFnOption::Kind("context_first".to_string())
        } else if lookahead.peek(kw::context_last) {
            XPathFnOption::Kind("context_last".to_string())
        } else if lookahead.peek(kw::position) {
            XPathFnOption::Kind("position".to_string())
        } else if lookahead.peek(kw::size) {
            XPathFnOption::Kind("size".to_string())
        } else if lookahead.peek(LitStr) {
            let string_literal: LitStr = input.parse()?;
            let signature = string_literal.value();
            XPathFnOption::Signature(signature)
        } else {
            bail_spanned!(
                input.span() => "Expected a string literal or a context keyword"
            );
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_parse_signature() {
        assert_debug_snapshot!(syn::parse_str::<XPathFnOptions>(
            r#""fn:foo() as xs:string""#
        ));
    }

    #[test]
    fn test_parse_signature_with_kind() {
        assert_debug_snapshot!(syn::parse_str::<XPathFnOptions>(
            r#""fn:foo() as xs:string", context_first"#
        ));
    }

    #[test]
    fn test_parse_signature_parse_error() {
        assert_debug_snapshot!(syn::parse_str::<XPathFnOptions>(r#""wrong wrong""#));
    }

    #[test]
    fn test_parse_not_a_signature_string() {
        assert_debug_snapshot!(syn::parse_str::<XPathFnOptions>(r#"wrong"#));
    }

    #[test]
    fn test_parse_signature_unknown_kind() {
        assert_debug_snapshot!(syn::parse_str::<XPathFnOptions>(
            r#""fn:foo() as xs:string",blah"#
        ));
    }
}
