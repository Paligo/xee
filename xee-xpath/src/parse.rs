// parse XPath via pest

// XXX extra-grammatical constraints still need to be handled:
// https://www.w3.org/TR/xpath-31/#extra-grammatical-constraints
// XXX reserved function names still need to be handled:
// https://www.w3.org/TR/xpath-31/#id-reserved-fn-names

#[derive(Parser)]
#[grammar = "xpath-31.pest"]
pub struct XPathParser;

#[cfg(test)]
mod tests {
    use super::*;
    use pest::Parser;

    #[test]
    fn test_char() {
        let successful_parse = XPathParser::parse(Rule::Char, "abc");
        assert!(successful_parse.is_ok());
        let unsuccessful_parse = XPathParser::parse(Rule::Char, "\u{10}");
        assert!(unsuccessful_parse.is_err());
    }

    #[test]
    fn test_xpath() {
        let successful_parse = XPathParser::parse(Rule::Xpath, "/a/c");
        assert!(successful_parse.is_ok());
        let unsuccessful_parse = XPathParser::parse(Rule::Xpath, ")(");
        assert!(unsuccessful_parse.is_err());
    }
}
