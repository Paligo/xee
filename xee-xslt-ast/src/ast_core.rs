use xee_xpath_ast::ast as xpath_ast;

type XPathExpr = xpath_ast::XPath;
type EqName = String;
type SequenceType = String;

#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct If {
    pub(crate) test: XPathExpr,
    pub(crate) content: Vec<SequenceConstructor>,
}

#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub enum SequenceConstructor {
    Text(String),
}

#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct Variable {
    pub name: EqName,
    // TODO: should this be subsumed into the
    // content, as a variable may either have
    // a select or a content, but not both
    pub select: Option<XPathExpr>,
    // as_: Option<SequenceType>,
    // static_: bool,
    // visbility: Visibility,
    // // it's also possible to have an empty variable
    // // in case visibility is static; this could
    // // perhaps be modelled separately?
    pub content: Vec<SequenceConstructor>,
}

enum Visibility {
    Public,
    Private,
    Final,
    Abstract,
}

struct Function {
    name: EqName,
    as_: Option<SequenceType>,
    visbility: Visibility,
    override_extension_function: bool,
    new_each_time: NewEachTime,
    cache: bool,
    params: Vec<Param>,
    content: Vec<SequenceConstructor>,
}

struct Param {}

enum NewEachTime {
    Bool(bool),
    Maybe,
}

#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub enum Instruction {
    Variable(Variable),
    If(If),
}
