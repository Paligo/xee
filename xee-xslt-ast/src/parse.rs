use ahash::HashMap;
use chumsky::util::MaybeRef;
use chumsky::{extra::Full, input::ValueInput, prelude::*};
use std::borrow::Cow;
use xee_xpath_ast::Namespaces;
use xot::{Node, Xot};

use crate::ast_core as ast;

pub(crate) struct State<'a> {
    pub(crate) namespaces: Cow<'a, Namespaces<'a>>,
}

type Extra<'a> = Full<ParserError, State<'a>, ()>;

pub(crate) type BoxedParser<'a, I, T> = Boxed<'a, 'a, I, T, Extra<'a>>;

pub type Span = SimpleSpan;

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Token<'a> {
    ElementStart(Name<'a>, HashMap<Name<'a>, &'a str>),
    ElementEnd(Name<'a>),
    Text(&'a str),
    Comment(&'a str),
    ProcessingInstruction(&'a str, Option<&'a str>),
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Name<'a> {
    namespace: &'a str,
    localname: &'a str,
}

impl<'a> From<(&'a str, &'a str)> for Name<'a> {
    fn from((localname, namespace): (&'a str, &'a str)) -> Self {
        Self {
            namespace,
            localname,
        }
    }
}

#[cfg_attr(test, derive(serde::Serialize))]
pub enum ParserError {
    ExpectedFound { span: Span },
    MyError,
    XPath(xee_xpath_ast::ParserError),
}

impl From<xee_xpath_ast::ParserError> for ParserError {
    fn from(e: xee_xpath_ast::ParserError) -> Self {
        Self::XPath(e)
    }
}

impl<'a, I> chumsky::error::Error<'a, I> for ParserError
where
    I: ValueInput<'a, Token = Token<'a>, Span = Span>,
{
    // we don't do anything with expected and found, instead just retaining
    // the span. This is because these contain tokens with a lifetime, and
    // having a lifetime for the ParserError turns out open up a world of trouble
    // as soon as we want to build on it in the XSLT parser. We also don't
    // have a good way to turn a logos token into a human-readable string, so
    // we couldn't really construct good error messages anyway.
    fn expected_found<E: IntoIterator<Item = Option<MaybeRef<'a, Token<'a>>>>>(
        _expected: E,
        _found: Option<MaybeRef<'a, Token<'a>>>,
        span: Span,
    ) -> Self {
        Self::ExpectedFound { span }
    }

    fn merge(self, other: Self) -> Self {
        match (self, other) {
            (
                ParserError::ExpectedFound { span: span_a },
                ParserError::ExpectedFound { span: _ },
            ) => ParserError::ExpectedFound { span: span_a },
            (ParserError::ExpectedFound { .. }, a) => a,
            (a, ParserError::ExpectedFound { .. }) => a,
            (a, _) => a,
        }
    }
}

struct TokenizedTraverse<'a, T: Iterator<Item = xot::NodeEdge>> {
    xot: &'a Xot,
    traverse: T,
}

impl<'a, T: Iterator<Item = xot::NodeEdge>> Iterator for TokenizedTraverse<'a, T> {
    type Item = Token<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        let edge = self.traverse.next()?;
        Some(match edge {
            xot::NodeEdge::Start(node) => match self.xot.value(node) {
                xot::Value::Element(e) => {
                    let name: Name = self.xot.name_ns_str(e.name()).into();
                    let attributes = e
                        .attributes()
                        .iter()
                        .map(|(name, value)| {
                            let name: Name = self.xot.name_ns_str(*name).into();
                            (name, value.as_ref())
                        })
                        .collect::<HashMap<_, _>>();
                    Token::ElementStart(name, attributes)
                }
                xot::Value::Text(text) => Token::Text(text.get()),
                xot::Value::Comment(comment) => Token::Comment(comment.get()),
                xot::Value::ProcessingInstruction(pi) => {
                    Token::ProcessingInstruction(pi.target(), pi.data())
                }
                xot::Value::Root => self.next()?,
            },
            xot::NodeEdge::End(node) => {
                if let xot::Value::Element(e) = self.xot.value(node) {
                    let name: Name = self.xot.name_ns_str(e.name()).into();
                    Token::ElementEnd(name)
                } else {
                    self.next()?
                }
            }
        })
    }
}

fn tokenize(xot: &Xot, node: Node) -> TokenizedTraverse<impl Iterator<Item = xot::NodeEdge> + '_> {
    TokenizedTraverse {
        xot,
        traverse: xot.traverse(node),
    }
}

fn parser<'a, I>() -> BoxedParser<'a, I, ast::If>
where
    I: ValueInput<'a, Token = Token<'a>, Span = Span>,
{
    let if_ = select! {
        Token::ElementStart(Name { namespace: "", localname: "if"}, attrs) => attrs,
    }
    .boxed();

    let text = select! {
        Token::Text(s) => s,
    }
    .boxed();

    let sequence_constructor = text
        .map(|text| ast::SequenceConstructorItem::Text(text.to_string()))
        .boxed();

    if_.then(sequence_constructor)
        .try_map_with_state(|(attributes, content), _span, state: &mut State| {
            let name = Name {
                namespace: "",
                localname: "test",
            };
            let test = attributes.get(&name).unwrap();
            let namespaces = state.namespaces.as_ref();

            let test = xee_xpath_ast::ast::XPath::parse(test, namespaces, &[])?;
            Ok(ast::If {
                test,
                content: vec![content],
            })
        })
        .then_ignore(just(Token::ElementEnd(Name {
            namespace: "",
            localname: "if",
        })))
        .boxed()
}

#[cfg(test)]
mod tests {
    use super::*;
    use chumsky::input::Stream;
    use insta::assert_ron_snapshot;

    #[test]
    fn test_tokenize() {
        let mut xot = Xot::new();
        let doc = xot.parse(r#"<if test="true()">Hello</if>"#).unwrap();
        let tokens = tokenize(&xot, doc).collect::<Vec<_>>();
        assert_ron_snapshot!(tokens);
    }

    #[test]
    fn test_simple_parse_if() {
        let mut xot = Xot::new();
        let doc = xot.parse(r#"<if test="true()">Hello</if>"#).unwrap();
        let tokens = tokenize(&xot, doc);
        let stream = Stream::from_iter(tokens);
        let namespaces = Namespaces::default();
        let mut state = State {
            namespaces: Cow::Owned(namespaces),
        };
        assert_ron_snapshot!(parser().parse_with_state(stream, &mut state).into_result());
    }
}
