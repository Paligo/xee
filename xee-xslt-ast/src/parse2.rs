use chumsky::util::MaybeRef;
use chumsky::{extra::Full, input::ValueInput, prelude::*};
use xmlparser::{StrSpan, Token};

use crate::parse::{Span, State};

use crate::ast_core as ast;

type Extra<'a> = Full<ParserError, State<'a>, ()>;

pub(crate) type BoxedParser<'a, I, T> = Boxed<'a, 'a, I, T, Extra<'a>>;

#[cfg_attr(test, derive(serde::Serialize))]
pub enum ParserError {
    ExpectedFound { span: Span },
    MyError,
    XPath(xee_xpath_ast::ParserError),
}

impl From<xee_xpath_ast::ParserError> for ParserError {
    fn from(e: xee_xpath_ast::ParserError) -> Self {
        Self::XPath(e)
    }
}

impl<'a, I> chumsky::error::Error<'a, I> for ParserError
where
    I: ValueInput<'a, Token = Token<'a>, Span = Span>,
{
    // we don't do anything with expected and found, instead just retaining
    // the span. This is because these contain tokens with a lifetime, and
    // having a lifetime for the ParserError turns out open up a world of trouble
    // as soon as we want to build on it in the XSLT parser. We also don't
    // have a good way to turn a logos token into a human-readable string, so
    // we couldn't really construct good error messages anyway.
    fn expected_found<E: IntoIterator<Item = Option<MaybeRef<'a, Token<'a>>>>>(
        _expected: E,
        _found: Option<MaybeRef<'a, Token<'a>>>,
        span: Span,
    ) -> Self {
        Self::ExpectedFound { span }
    }

    fn merge(self, other: Self) -> Self {
        match (self, other) {
            (
                ParserError::ExpectedFound { span: span_a },
                ParserError::ExpectedFound { span: _ },
            ) => ParserError::ExpectedFound { span: span_a },
            (ParserError::ExpectedFound { .. }, a) => a,
            (a, ParserError::ExpectedFound { .. }) => a,
            (a, _) => a,
        }
    }
}

struct Element<'a>(&'a str);

struct Attribute<'a>(&'a str, &'a str);

fn element_start<'a, I>(local: &'a str) -> BoxedParser<'a, I, ()>
where
    I: ValueInput<'a, Token = Token<'a>, Span = Span>,
{
    let element = select! {
        Token::ElementStart { local, .. } => Element(local.as_str()),
    }
    .boxed();
    element
        .validate(move |element, span, emitter| {
            if element.0 != local {
                emitter.emit(ParserError::ExpectedFound { span })
            }
        })
        .boxed()
}

fn element_end<'a, I>() -> BoxedParser<'a, I, ()>
where
    I: ValueInput<'a, Token = Token<'a>, Span = Span>,
{
    select! {
        Token::ElementEnd { end: xmlparser::ElementEnd::Open, .. } => (),
    }
    .boxed()
}

fn element_close<'a, I>() -> BoxedParser<'a, I, ()>
where
    I: ValueInput<'a, Token = Token<'a>, Span = Span>,
{
    select! {
        Token::ElementEnd { end: xmlparser::ElementEnd::Close(..) | xmlparser::ElementEnd::Empty, .. } => (),
    }.boxed()
}

fn parser<'a, I>() -> BoxedParser<'a, I, ast::If>
where
    I: ValueInput<'a, Token = Token<'a>, Span = Span>,
{
    let if_ = element_start("if");
    // let element = select! {
    //     Token::ElementStart { local, .. } => Element(local.as_str()),
    // }
    // .boxed();
    // let if_ = element
    //     .validate(|element, span, emitter| {
    //         if element.0 != "if" {
    //             emitter.emit(ParserError::ExpectedFound { span })
    //         }
    //     })
    //     .boxed();
    // let element_end = select! {
    //     Token::ElementEnd { end: xmlparser::ElementEnd::Open, .. } => (),
    // }
    // .boxed();

    // let element_close = select! {
    //     Token::ElementEnd { end: xmlparser::ElementEnd::Close(..) | xmlparser::ElementEnd::Empty, .. } => (),
    // };

    let text = select! {
        Token::Text { text } => text,
    }
    .boxed();

    let sequence_constructor = text
        .map(|text| ast::SequenceConstructor::Text(text.to_string()))
        .boxed();

    let attribute = select! {
        Token::Attribute { local, value, ..} => Attribute(local.as_str(), value.as_str()),
    }
    .boxed();

    let test_attribute_str = attribute
        .validate(|attribute, span, emitter| {
            if attribute.0 != "test" {
                emitter.emit(ParserError::ExpectedFound { span })
            }
            attribute.1
        })
        .boxed();

    let test_attribute =
        test_attribute_str.try_map_with_state(|value, _span, state: &mut State| {
            Ok(xee_xpath_ast::ast::XPath::parse(
                value,
                state.namespaces.as_ref(),
                &[],
            )?)
        });

    let if_attributes = test_attribute.repeated().collect::<Vec<_>>();

    if_.ignore_then(if_attributes)
        .then_ignore(element_end())
        .then(sequence_constructor)
        .try_map_with_state(|(attributes, content), _span, _state: &mut State| {
            let test = attributes.into_iter().next().unwrap();
            Ok(ast::If {
                test,
                content: vec![content],
            })
        })
        .then_ignore(element_close())
        .boxed()
}

#[cfg(test)]
mod tests {
    use std::borrow::Cow;

    use super::*;
    use chumsky::input::Stream;
    use insta::assert_ron_snapshot;
    use xee_xpath_ast::Namespaces;

    // #[test]
    // fn test_tokens() {
    //     let tokens = xmlparser::Tokenizer::from(r#"<if test="true()">Hello</if>"#);

    //     dbg!(tokens.collect::<Vec<_>>());
    // }

    #[test]
    fn test_simple_parse_if() {
        let tokens = xmlparser::Tokenizer::from(r#"<if test="true()">Hello</if>"#);

        let stream = Stream::from_iter(tokens.map(|t| t.unwrap()));
        let namespaces = Namespaces::default();
        let mut state = State {
            namespaces: Cow::Owned(namespaces),
        };
        assert_ron_snapshot!(parser().parse_with_state(stream, &mut state).into_result());
    }
}
