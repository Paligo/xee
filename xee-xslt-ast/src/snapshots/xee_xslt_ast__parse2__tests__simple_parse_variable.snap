---
source: xee-xslt-ast/src/parse2.rs
expression: "parser().parse_with_state(stream, &mut state).into_result()"
---
Ok(Variable(Variable(
  name: "foo",
  select: Some(XPath(Expr([
    Path(PathExpr(
      steps: [
        PrimaryExpr(FunctionCall(FunctionCall(
          name: Name(
            name: "true",
            prefix: None,
            namespace: Some("http://www.w3.org/2005/xpath-functions"),
          ),
          arguments: [],
        ))),
      ],
    )),
  ]))),
  content: [
    Text("Hello"),
  ],
)))
